Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+import Logging
+import SwiftProtobuf
+
+enum RecieveResponseHeadError: Error, Equatable {
+  /// The 'content-type' header was missing or the value is not supported by this implementation.
+  case invalidContentType
+
+  /// The HTTP response status from the server was not 200 OK.
+  case invalidHTTPStatus(HTTPResponseStatus)
+
+  /// The encoding used by the server is not supported.
+  case unsupportedMessageEncoding
+
+  /// An invalid state was encountered. This is a serious implementation error.
+  case invalidState
+}
+
+enum SendEndOfRequestStreamError: Error {
+  /// The request stream has already been closed.
+  case endOfStreamCardinalityViolation
+
+  /// An invalid state was encountered. This is a serious implementation error.
+  case invalidState
+}
+
+  /// An invalid state was encountered. This is a serious implementation error.
+struct InvalidStateError: Error {
+  let function: String
+  static func invalidState(_ function: String = #function) -> InvalidStateError {
+    return .init(function: function)
+  }
+}
+
+/// A state machine for a single gRPC call from the perspective of a client.
+///
+/// See: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+struct GRPCClientStateMachine<Request: Message, Response: Message> {
+  /// The combined state of the request (client) and response (server) streams for an RPC call.
+  ///
+  /// The following states are not possible:
+  /// - `.clientIdleServerStreaming`: The client must initiate the call before the server moves
+  ///   from the idle state.
+  /// - `.clientIdleServerClosed`: The client must initiate the call before the server moves from
+  ///   the idle state.
+  /// - `.clientStreamingServerClosed`: The client may not stream if the server is closed.
+  enum State {
+    /// Initial state. Neither request stream nor response stream have been initiated. Holds the
+    /// pending write state for the request stream and expected message count for the response
+    /// stream, respectively.
+    ///
+    /// Valid transitions:
+    /// - `clientStreamingServerIdle`: if the client initiates the RPC,
+    /// - `clientClosedServerClosed`: if the client terminates the RPC.
+    case clientIdleServerIdle(PendingWriteState, MessageCount)
```

Comment:
Usually I don't like them because the labels tend to just match the types so you end up with e.g.:

```swift
switch self {
  case .clientIdleServerIdle(pendingWriteState: let pendingWriteState, messageCount: let messageCount):
    // ...
}
```

Perhaps `client` and `server` labels would be useful though?
```swift
switch self {
  case .clientIdleServerIdle(client: let pendingWriteState, server: let messageCount):
    // ...
}
```

Reply:
I think you can actually omit the label (e.g. `pendingWriteState: `) in `case` statements, could you check that?